### ISSUE TITLE
Maybe warn for sketchy null checks on nullable enum types?

### ISSUE THREAD

**Issue by oyamauchi** [facebook/hhvm#4963]
Right now, we warn if you use equality-comparison or bool-cast to null-check a value of nullable primitive type:

```hack
function nullable_int(): ?int { return null; }
function main() {
  $i = nullable_int();
  if ($i) {  // Warning: You are using a sketchy null check on a primitive value
    // ...
  }
}
```

However, we don't do the same for enums. Shouldn't we actually do that, since enums are actually primitive types too?

```hack
enum E: int {
  VAL = 0;
}
function main() {
  $x = 0;
  $e = E::coerce($x);
  // $e is now of type ?E
  if ($e) {  // No warning, but this isn't good...
    // $e is now of type E
  }
}
```

One other thing I noticed is that there's a version of the "sketchy null check" warning that doesn't talk about primitive types. However, I can't see a way to trigger it. It seems to only apply if you're null-checking a value of type "nullable `mixed`" or "nullable `Tany`", neither of which is possible to get now. The relevant code is here: https://github.com/facebook/hhvm/blob/master/hphp/hack/src/typing/typing.ml#L3278

**Comment by jwatzman** [comment]
Yeah I agree we should add this check for first-class enums (prob an oversight that it doesn't work now), as well as kill off the vestigial check you point to. ([Link including git hash in case the file changes in the future](https://github.com/facebook/hhvm/blob/c3140308824f7eb7301b65179545623e210dd390/hphp/hack/src/typing/typing.ml#L3278))

**Comment by oyamauchi** [comment]
I'm gonna take a crack at it, then.

**Comment by oyamauchi** [comment]
After some internal discussion, we decided we're not going to do this.

If we just naïvely warn on every sketchy null check of an enum, most of them turn out to be false positives, since in practice most enums don't have falsey values (at FB anyway; most enums are strings). I considered restricting the checks to enums that have falsey values, but determining what the values actually are in enums is nontrivial. hh_server can't evaluate an arbitrary expression, or even just the ones that are valid on the RHS of an enum member definition. Adding that ability seems like overkill for this purpose. A compromise solution might be to make a best effort: only look at scalar expressions on the RHS of member definitions, and warn based on that, but that seems excessively hacky to me.

All in all, we've convinced ourselves it's not that bad to let this problem go un-warned; or at least, it's not as bad as the possible solutions.
